Adding dev server overrides

cereallarceny opened the pull request:
I added the ability to add overrides to the default dev server configuration. This should only be merged after https://github.com/yusinto/universal-hot-reload/pull/31

cereallarceny commented:
@yusinto I've updated this PR to have all the changes from #31 and now is much smaller. This simply adds the ability to all the dev server to be overridden. I've already confirmed this works on my own local setup and have written tests for this. I believe it's very safe to merge and you can continue closing all other PR's after merging this one.

cereallarceny commented:
@yusinto I have also now repaired the build from linting discrepancies. They should now be fixed in this PR.

yusinto commented on src/client/watchClientChanges.js:
```diff
@@ -5,7 +5,7 @@ import webpackDevServer from 'webpack-dev-server';
  * Start webpack dev server for hmr
  */
-const watchClientChanges = clientConfig => {
+const watchClientChanges = (clientConfig, devServerOverrides, verbose) => {
```
A couple of things here. Webpack already supports a [`devServer` option](https://webpack.js.org/configuration/dev-server/#devserver) through the standard webpack config. We should be able to use this instead of having to add our own overrides. 

Secondly re verbose logging, I think we can have a better way to control logging output than just placing a conditional around every console logs. I've done this [here](https://github.com/yusinto/pagerduty-webhook-processor/blob/master/src/logger.js) with my own logger class, but the idea is to be able to control log output from a central place. I think this should be a separate PR though, because it's unrelated to the your original intention of this PR which is to be able to customise dev server options.

cereallarceny replied:
@yusinto Unless I misunderstand your comment, I don't think you understand what I'm trying to do here. The `devServerOverrides` is passed into your webpack dev server configuration that you have set up on the client. It allows a user of universal-hot-reload the ability to override the default configuration - you can see that here: https://github.com/yusinto/universal-hot-reload/pull/32/files#diff-799424c7a9cd2ab633aefd9fb1b04240R32-R44

For instance, if I want to turn on verbose stats instead of leaving them at the default of `errors-only` this `devServerOverrides` variable allows me to do that. Look at the whole PR in context and I think this should make a bit more sense.

cereallarceny replied:
To give an example, I can now do something like:

```
UniversalHotReload({
  clientConfig,
  serverConfig,
  devServerOptions: {
    stats: 'verbose'
  }
});
```

If you look at the tests I wrote for this PR they show this capability and demonstrate that it's working as intended.

yusinto replied:
@cereallarceny I understand clearly what you are trying to achieve. My suggestion was that consumer can set `clientConfig.devServer` and then we can use that object instead of having to specify a new param called `devServerOptions`. 

```tsx
UniversalHotReload({
  clientConfig, // this contains "devServer" object
  serverConfig,
  devServerOptions: {
    stats: 'verbose'
  }
});
```

Then in `watchClientChanges.js` we can merge this object with the hardcoded defaults `devServerOptions`. Does this make sense?

yusinto reviewed (changes requested):
Hi @cereallarceny thanks for this, I've left a comment here so please take a look. Thank you.

cereallarceny commented:
@yusinto I'm happy for us to use your logger. It doesn't matter to me what logging class we use - but I figured mine would be good enough for most use-cases. I just want to turn off your logs. If you want to add a finer level of control, then you can do so in a future PR. For now though, I think this offers reasonable functionality.

yusinto commented:
> If you want to add a finer level of control, then you can do so in a future PR.

I'm happy to do the custom logger bit in the future. In which case for this PR, perhaps you can omit the `verbose` options for now until I have that in place. This does mean you'll have to put up with the console logs until my pr comes around though, so apologies if advance for that if this bothers you tremendously.

cereallarceny commented:
@yusinto Yes, that does disappoint me. Silencing all the logs of your library's output is the final thing I need to be able to use this library in production. The lack of this feature is preventing me from launching my own library. I hope you can address this in your own PR soon - otherwise, I'm going to fork this project and make my own changes. I'm just trying to help here. 😄 

Either way, I've changed your `devServer` config to be passed the way you want - you're right, much cleaner! I've also removed the verbose flag so that you may do your PR.

Also, your test suite wasn't passing. I wasn't sure quite why so I fixed a bunch of unrelated problems with the test suite and everything is correct now. Looks like some typos. You can merge now, I've done everything you requested.

cereallarceny commented:
@yusinto Can we please get this merged? I'm going to need to fork and separately release my own version of this library if we can't hurry along merging pull requests.

yusinto commented:
@cereallarceny thanks for your work! I'll create a new release today. Thanks again for helping out.
